client/asset/dcr/externaltx.go

```diff
@@ -0,0 +1,475 @@
+	spent := spenderTx != nil
+	if spent {
+		spenderBlockHash, err := chainhash.NewHashFromStr(spenderTx.BlockHash)
+		if err != nil {
+			// dcr.log.Errorf("Invalid hash (%s) for tx that spends output %s: %v", spenderTx.BlockHash, output.op, err)
+			return false, fmt.Errorf("invalid hash (%s) for tx that spends output %s: %v", spenderTx.BlockHash, output.op, err)
+		} else {
+			output.spenderBlock = &block{hash: spenderBlockHash, height: spenderTx.BlockHeight}
+		}
+	}
+	return spent, err
+}
+
+// findTxOutSpender attempts to find and return the tx that spends the provided
+// output by matching the provided outputPkScript against the block filters of
+// the mainchain blocks between the provided startBlock and the current best
+// block.
+// If no tx is found to spend the provided output, the hash of the block that
+// was last checked is returned along with any error that may have occurred
+// during the search.
+func (dcr *ExchangeWallet) findTxOutSpender(ctx context.Context, op outPoint, outputPkScript []byte, startBlock *block) (*chainjson.TxRawResult, *chainhash.Hash, error) {
+	iHeight := startBlock.height
+	iHash := startBlock.hash
+	bestBlock := dcr.cachedBestBlock()
+	for {
+		blockFilter, err := dcr.getBlockFilterV2(ctx, iHash)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if blockFilter.Match(outputPkScript) {
+			dcr.log.Debugf("Output %s is likely spent in block %d (%s). Confirming.",
+				op, iHeight, iHash)
+			blk, err := dcr.getBlock(ctx, iHash, true)
+			if err != nil {
+				return nil, iHash, err
+			}
+			blockTxs := append(blk.RawTx, blk.RawSTx...)
+			for i := range blockTxs {
+				blkTx := &blockTxs[i]
+				if txSpendsOutput(blkTx, op) {
+					dcr.log.Debugf("Found spender for output %s in block %d (%s), spender tx hash %s.",
+						op, iHeight, iHash, blkTx.Txid)
+					return blkTx, iHash, nil
+				}
+			}
+			dcr.log.Debugf("Output %s is NOT spent in block %d (%s).", op, iHeight, iHash)
+		}
+
+		if iHeight >= bestBlock.height { // reached the tip, stop searching
+			break
+		}
+
+		// Block does not include the output spender, check the next block.
+		iHeight++
+		nextHash, err := dcr.wallet.GetBlockHash(ctx, iHeight)
+		if err != nil {
+			return nil, iHash, translateRPCCancelErr(err)
+		}
+		iHash = nextHash
+	}
+
+	dcr.log.Debugf("Output %s is NOT spent in blocks %d (%s) to %d (%s).",
+		op, startBlock.height, startBlock.hash, bestBlock.height, bestBlock.hash)
+	return nil, bestBlock.hash, nil // scanned up to best block, no spender found
+}
+
+// txSpendsOutput returns true if the passed tx has an input that spends the
+// specified output.
+func txSpendsOutput(tx *chainjson.TxRawResult, op outPoint) bool {
+	if tx.Txid == op.txHash.String() {
+		return false // no need to check inputs if this tx is the same tx that pays to the specified op
+	}
+	for i := range tx.Vin {
+		input := &tx.Vin[i]
+		if input.Vout == op.vout && input.Txid == op.txHash.String() {
```

Comment:
`String` will be called multiple times, starting with the initial `tx.Txid` check, so might as well make a local var on the first line of the func.